Create adopting-prebuild.md (#18613)
expo/expo: docs/pages/guides/adopting-prebuild.md

---
title: Adopting Prebuild
---

import { Terminal } from '~/ui/components/Snippet';

> There are many advantages to using Prebuild to manage your native projects. This guide will show you how to adopt prebuilding in a project that was bootstrapped with `npx react-native init`.

Reviewer: ```suggestion
There are many advantages of using Prebuild to manage your native projects. This guide will show you how to adopt prebuilding in a project that was bootstrapped with `npx react-native init`.
```

> **Note:** Adopting prebuild automatically adds support for developing modules with the [Sweet API][expo-modules-core] by linking `expo-modules-core` natively. You can also use any command from Expo CLI `npx expo` in your project.

## Bootstrap a new project

<!-- NOTE: Update the version when we bump support -->

<Terminal cmd={["$ npx react-native init --version 0.69.4"]} />

> This will create a new React Native project. Note that not all versions of `react-native` are supported, be sure to use a version of `react-native` that has a corresponding Expo SDK version.

Reviewer 2: ```suggestion
This will create a new React Native project. Note that [not all versions of `react-native` are explicitly supported](/versions/latest/#each-expo-sdk-version-depends-on-a), be sure to use a version of `react-native` that has a corresponding Expo SDK version.
```

## Install the `expo` package

The `expo` package contains the `npx expo prebuild` command and indicates which prebuild template to use:

<Terminal cmd={["$ yarn add expo"]} />

Ensure you install the version of `expo` that works for your currently installed version of `react-native`. Note that not all major versions of `react-native` are supported.

## Update the entry file

Modify the entry file like so:

```diff
+ import {registerRootComponent} from 'expo';

- import {AppRegistry} from 'react-native';
import App from './App';
- import {name as appName} from './app.json';

- AppRegistry.registerComponent(appName, () => App);
+ registerRootComponent(App);
```

> Alternatively, you can just change `appName` to `"main"`. Switching to `registerRootComponent` is useful if you want to take advantage of other features like web support. Importing `expo` adds side-effects for development environments that are removed in production bundles, meaning there isn't some massive bundle size increase.

Reviewer 2: ```suggestion
Alternatively, you can change `appName` to `"main"`. Switching to `registerRootComponent` is useful if you want to take advantage of other features like web support. Importing `expo` adds side-effects for development environments that are removed in production bundles.
```

> ## Prebuild

Reviewer 3: Why would you manually want to run prebuild at this point? If the npm scripts are updated to call `expo run:...` below, then wouldn't it be much easier to just say: 'If you haven't made any native file modifications in your project, it's safe to delete the ios and android folders at this point. They will be generated automatically for you when you start your project`

Reviewer 2: my take is that it's just more explicit to run prebuild here, and the app might not actually work yet if you were to go straight to `expo run:ios` or `expo run:android`

]} cmdCopy="npx expo prebuild --clean" />

You can test that everything worked by building the projects locally:

<Terminal cmd={[
'# Build your native Android project',
'$ npx expo run:android',
'',
'# Build your native iOS project',
'$ npx expo run:ios'
]} />

## Extra changes

The following changes are optional but recommended.

**.gitignore**

You can add `.expo/` to your `.gitignore` to prevent generated values from Expo CLI from being committed. You can also add `ios/` and `android/` to the `.gitignore` if you want to ensure they aren't committed between prebuilds.

**app.json**

Remove all fields that are outside of the top-level `expo` object as these will not be used in `npx expo prebuild`.

```diff
{
-  "name": "myapp",
-  "displayName": "myapp"
+  "expo": {
+    "name": "myapp"
+  }
}
```

**metro.config.js**

- See [Customizing Metro](/guides/customizing-metro.md)

**package.json**

You may want to change the scripts to use the Expo CLI run commands:

```diff
  "scripts": {
    "start": "expo start --dev-client",
-    "android": "react-native run-android",
-    "ios": "react-native run-ios",
+    "android": "expo run:android",
+    "ios": "expo run:ios",
  },
```

> 

Reviewer 2: should we add babel config too?

Author: Babel config is required for web support and OTA updates, it doesn't overlap with `prebuild` afaict.

These have better logging, auto code signing, better simulator handling, and they ensure you run `npx expo start` to host files.

> ## Advanced Migration

Reviewer 2: ```suggestion
## Re-applying native customizations
```

> If you're project already has a number of custom native modifications then you'll need to configure your Expo config (`app.json`) to reflect those native changes. There is currently no automated system for doing this. So instead we'll provide some pointers:

Reviewer 2: i think this needs to be above any optional parts. this is going to be necessary for any but the most trivial of projects

Author: This section isn't very concrete and generally just provides tips for migration which is why I made it more of a secondary guide and originally had it titled as such.

> If you're project already has a number of custom native modifications then you'll need to configure your Expo config (`app.json`) to reflect those native changes. There is currently no automated system for doing this. So instead we'll provide some pointers:

Reviewer 2: ```suggestion
If your project has any native modifications (changes to the `ios` or `android` directories, such as app icon configuration or splash screen) then you'll need to configure your Expo config (`app.json`) to reflect those native changes.
```

- Check to see if your changes overlap with the built-in [Expo config fields](/versions/latest/config/app/). For example, if you have an app icon, be sure to define it as `expo.icon` in the `app.json` then re-run `npx expo prebuild`.
- Look up if any of the packages you're using require an [Expo Config Plugin][config-plugins]. If a package in your project required additional changes inside of the `ios/` or `android/` folders, then you will probably need a Config Plugin. Some plugins can be automatically added by running `npx expo install` with all of the packages in your `package.json` `dependencies`. If a package requires a plugin but doesn't supply one, then you can try checking the community plugins at [`expo/config-plugins`](https://github.com/expo/config-plugins) to see if one already exists.

> - You can use the [VS code Expo extension][vs-code-expo] to "introspect" your changes and debug if prebuild is generating the native code you expect. Just press `cmd+shift+P`, type "Expo: Preview Modifier", and select the native file you wish to introspect.

Reviewer: ```suggestion
- You can use the [VS Code Expo extension][vs-code-expo] to introspect your changes and debug if prebuild is generating the native code you expect. Just press `cmd+shift+P`, type "Expo: Preview Modifier", and select the native file you wish to introspect.
```

Let's not use quotes to highlight a text or a word. If you think it's important to highlight something, let's bold text.

- Additionally, you can develop local Config Plugins to fit your needs. [Learn more](/guides/config-plugins#developing-a-plugin).

## Adding More Features

Prebuild is the tip of the automation iceberg, here are some features you can adopt next:

- [EAS Build](/build/setup): Code signing and cloud building.
- [EAS Update](/build/updates): Send over-the-air updates instantly.
- [Expo for web](/workflow/web): Run your app in the browser.

> - [Expo Dev Client][dev-client]: Improved developer experience for teams.

Reviewer 2: ```suggestion
- [Expo Dev Client][dev-client]: Create your own personal "Expo Go" type app around your native runtime.
```

> - [Sweet API][expo-modules-core]: Write modules with Swift and Kotlin. This is automatically supported when using `npx expo prebuild`.

Author: ```suggestion
- [Expo native module API][expo-modules-core]: Write modules with Swift and Kotlin. This is automatically supported when using `npx expo prebuild`.
```
